### PR TITLE
[libc] Add more functions in CPP/bit.h

### PR DESCRIPTION
**Pull request by gchatelet**
Once this is submitted we can remove `include/__support/bit.h` that duplicates
some of this functionnality.

### CODE REVIEW

**Comment by SchrodingerZhu** [comment]
lgtm. A latter patch should migrate those from `__support/bit.h` to `CPP/bit.h`?

File: libc/test/src/__support/CPP/bit_test.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  EXPECT_EQ(uint16_t(4), bit_ceil(uint16_t(3)));
+  EXPECT_EQ(uint32_t(4), bit_ceil(uint32_t(3)));
+  EXPECT_EQ(uint64_t(4), bit_ceil(uint64_t(3)));
+
+  EXPECT_EQ(uint8_t(4), bit_ceil(uint8_t(4)));
+  EXPECT_EQ(uint16_t(4), bit_ceil(uint16_t(4)));
+  EXPECT_EQ(uint32_t(4), bit_ceil(uint32_t(4)));
+  EXPECT_EQ(uint64_t(4), bit_ceil(uint64_t(4)));
+
+  // The result is the largest representable value for each type.
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/bit_test.cpp:89]: "representable power of 2". The largest value would be 0xff...ff

File: libc/src/__support/CPP/bit.h [hunk 2]
```diff
@@ -1,57 +1,261 @@
...
+
+/// Count number of 0's from the least significant bit to the most
+///   stopping at the first 1.
+///
+/// Only unsigned integral types are allowed.
+///
+/// Returns cpp::numeric_limits<T>::digits on an input of 0.
+template <typename T>
+[[nodiscard]] LIBC_INLINE constexpr int countr_zero(T value) {
+  static_assert(cpp::is_unsigned_v<T>);
```
Comment (legrosbuffle) [inline libc/src/__support/CPP/bit.h:74]: Why is this not SFINAEd in ? "This overload participates in overload resolution only if T is an unsigned integer type"

File: libc/src/__support/CPP/bit.h [hunk 3]
```diff
@@ -1,57 +1,261 @@
...
+/// Count the number of ones from the most significant bit to the first
+/// zero bit.
+///
+/// Ex. countl_one(0xFF0FFF00) == 8.
+/// Only unsigned integral types are allowed.
+///
+/// Returns cpp::numeric_limits<T>::digits on an input of all ones.
+template <typename T>
+[[nodiscard]] LIBC_INLINE constexpr int countl_one(T value) {
+  static_assert(cpp::is_unsigned_v<T>);
```
Comment (legrosbuffle) [inline libc/src/__support/CPP/bit.h:152]: ditto (here and below)

File: libc/src/__support/CPP/limits.h [hunk 4]
```diff
@@ -36,41 +36,53 @@ template <> class numeric_limits<int> {
 public:
   LIBC_INLINE static constexpr int max() { return INT_MAX; }
   LIBC_INLINE static constexpr int min() { return INT_MIN; }
+  LIBC_INLINE_VAR static constexpr int digits = CHAR_BIT * sizeof(int) - 1;
```
Comment (legrosbuffle) [inline libc/src/__support/CPP/limits.h:39]: Refactor all these to a `consteval int digits_for<T>()` ? Or maybe even:

```
template <> class numeric_limits<int> : public numeric_limits_impl<int, INT_MAX, INT_MIN> {};
```

and derive everything in `numeric_limits_impl` ?

File: libc/src/__support/CPP/bit.h [hunk 5]
```diff
@@ -1,57 +1,261 @@
-//===-- Freestanding version of bit_cast  -----------------------*- C++ -*-===//
+//===-- Implementation of the C++20 bit header  -----------------*- C++ -*-===//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
+// This is inspired from LLVM ADT/bit.h header.
```
Comment (michaelrj-google) [inline libc/src/__support/CPP/bit.h:8]: nit: "inspired by"

File: libc/src/__support/CPP/bit.h [hunk 6]
```diff
@@ -1,57 +1,261 @@
...
+namespace LIBC_NAMESPACE::cpp {
 
 #if LIBC_HAS_BUILTIN(__builtin_memcpy_inline)
 #define LLVM_LIBC_HAS_BUILTIN_MEMCPY_INLINE
 #endif
 
-// This function guarantees the bitcast to be optimized away by the compiler for
-// GCC >= 8 and Clang >= 6.
-template <class To, class From>
+// This implementation of bit_cast requires trivially-constructible To, to avoid
```
Comment (michaelrj-google) [inline libc/src/__support/CPP/bit.h:27]: does this change cause any issues with existing bit_cast calls?

File: libc/src/__support/CPP/bit.h [hunk 7]
```diff
@@ -1,57 +1,261 @@
...
+  constexpr unsigned N = cpp::numeric_limits<T>::digits;
+  rotate = rotate % N;
+  if (!rotate)
+    return value;
+  if (rotate < 0)
+    return cpp::rotl(value, -rotate);
+  return (value >> rotate) | (value << (N - rotate));
 }
 
+// TODO: Do we need this function at all? How is it different from
```
Comment (michaelrj-google) [inline libc/src/__support/CPP/bit.h:257]: It looks like it's used in the syscall implementation. From what I can tell the goal is to ensure that pointers or floating point values are passed safely while also casting correctly for types smaller than long.

**Review by michaelrj-google — COMMENTED** [review]
The implementations LGTM, but do you have plans to use all of these functions?  If they don't have a use case right now they seem like they'd be easy enough to add later.

File: libc/src/__support/CPP/bit.h [hunk 8]
```diff
@@ -1,57 +1,261 @@
...
+  static_assert(cpp::is_unsigned_v<T>);
+  if (value < 2)
+    return 1;
+  return T(1) << cpp::bit_width<T>(value - 1u);
+}
+
+/// Count the number of set bits in a value.
+/// Ex. popcount(0xF000F000) = 8
+/// Returns 0 if the word is zero.
+template <typename T, typename = cpp::enable_if_t<cpp::is_unsigned_v<T>>>
```
Comment (SchrodingerZhu) [inline libc/src/__support/CPP/bit.h:207]: Perhaps an reference to https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel ? Or a few lines explanation like "counting bit set by dividing the integer into multiple lanes and reduce the result".

File: libc/src/__support/CPP/bit.h [hunk 9]
```diff
@@ -1,57 +1,261 @@
...
+
+template <typename T, typename = cpp::enable_if_t<cpp::is_unsigned_v<T>>>
+[[nodiscard]] LIBC_INLINE constexpr T rotl(T value, int rotate) {
+  constexpr unsigned N = cpp::numeric_limits<T>::digits;
+  rotate = rotate % N;
+  if (!rotate)
+    return value;
+  if (rotate < 0)
+    return cpp::rotr(value, -rotate);
+  return (value << rotate) | (value >> (N - rotate));
```
Comment (SchrodingerZhu) [inline libc/src/__support/CPP/bit.h:243]: @nickdesaulniers recommended a citation to "Safe, Efficient, and Portable Rotate in C/C++"
  // https://blog.regehr.org/archives/1063 in previous `__support/bit.h`. Maybe move it here?

**Comment by SchrodingerZhu** [comment]
> The implementations LGTM, but do you have plans to use all of these functions?  If they don't have a use case right now they seem like they'd be easy enough to add later.

I think popcount is used in some tests, rotl is used in hash; all previous use to clz/ctz can be changed to using CPP/bit.h instead as it is more close to cpp's std.

File: libc/src/__support/CPP/limits.h [hunk 10]
```diff
@@ -24,77 +25,72 @@ constexpr long long LLONG_MIN = 1LL << (LLONG_BIT_WIDTH - 1);
 constexpr unsigned long long ULLONG_MAX = ~0ULL;
 #endif
 
-template <class T> class numeric_limits {
```
[the remark this answers is not in the document]
Comment (gchatelet) [reply libc/src/__support/CPP/limits.h:27]: Done in https://github.com/llvm/llvm-project/pull/73926

**Pull request merged**